reco_utils/dataset/pandas_df_utils.py

```diff
@@ -316,54 +316,40 @@ def negative_feedback_sampler(
         3   1   0
     """
     # Get all of the users and items.
-    users = df[col_user].unique()
     items = df[col_item].unique()
-
-    # Create a dataframe for all user-item pairs
-    df_neg = user_item_pairs(
-        pd.DataFrame(users, columns=[col_user]),
-        pd.DataFrame(items, columns=[col_item]),
-        col_user,
-        col_item,
-        user_item_filter_df=df,
-    )
-    df_neg[col_label] = 0
-
-    df_pos = df.copy()
-    df_pos[col_label] = 1
-
-    df_all = pd.concat([df_pos, df_neg], ignore_index=True, sort=True)
-    df_all = df_all[[col_user, col_item, col_label]]
-
-    # Sample negative feedback from the combined dataframe.
-    df_sample = (
-        df_all.groupby(col_user)
-        .apply(
-            lambda x: pd.concat(
-                [
-                    x[x[col_label] == 1],
-                    x[x[col_label] == 0].sample(
-                        min(
-                            max(
-                                round(len(x[x[col_label] == 1]) * ratio_neg_per_user), 1
-                            ),
-                            len(x[x[col_label] == 0]),
-                        ),
-                        random_state=seed,
-                        replace=False,
-                    )
-                ],
-                ignore_index=True,
-                sort=True,
+    rng = np.random.default_rng(seed=seed)
+
+    def sample_items(user_df):
+        # Sample negative items for the data frame restricted to a specific user
+        unseen_items = np.setdiff1d(items, user_df[col_item])     
```

Comment:
This can be more efficient (you don't need to compute the set difference). I will modify it on Monday.